Refactor serialization of benchmarks

syl20bnr opened the pull request:
See changes below for a list of changes.

I  will add an "upload" function to  the persistence module in a further PR.

The current naming of the saved files need a bit of work, currently I just do `benchmarks_<timestamp>.json` but we could  have some race  condition and get some files overridden. --> We updated it to `bench_<name>_<timestamp>`.

The computed  values are saved in seconds, maybe we should  save them in milliseconds ? --> We  changed to microseconds.

You can see an example of produced file here: https://gist.github.com/syl20bnr/b0fdbbf0c3877b9e4ddf262a9cc388ac

### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

- https://github.com/tracel-ai/burn/issues/1072

### Changes

* flatten benchmarks data to make it easier to save documents to a database and query them
* split some information into their own fields like backend and device
* add new serialized info:
  - computed values (mean, median, variance, min, max)
  - number of samples
  - operation name
  - tensor shapes if any
* serialize to separate files, one file per benchmark run
* simplify persistence module to only a save method
* remove remaining `num_repeats` in benches execute functions. 

### Testing

Tested serialization to disk with all the available benchmarks.

syl20bnr commented:
For the `.expect()` messages, I am not sure what to put in them. I see a lot of people putting an explanation of the error there and also sometimes people putting what is expected to  work. The word "expect" suggests the latter but that's not what I see online most of the time 🤷🏻‍♂️  What's our code guideline on this ?

nathanielsimard commented on backend-comparison/benches/binary.rs:
```diff
@@ -15,6 +15,14 @@ impl<B: Backend, const D: usize> Benchmark for BinaryBenchmark<B, D> {
         "Binary Ops".into()
     }
 
+    fn operation(&self) -> Option<String> {
+        Some("binary".to_string())
+    }
```
I don't think we should have the name and the operation, the name is sufficient and more general than the operation.

syl20bnr replied:
I was  thinking about how we will query the database, I would rather have an operator field than using string search in a "free form" name (with 5 bench we have already inconsistency).

We could let the name be the free form field for nice naming and add more precise fields.

nathanielsimard commented on backend-comparison/benches/binary.rs:
```diff
@@ -15,6 +15,14 @@ impl<B: Backend, const D: usize> Benchmark for BinaryBenchmark<B, D> {
+        Some("binary".to_string())
+    }
+
+    fn shapes(&self) -> Option<Vec<String>> {
+        Some(vec![format!("{:?}", self.shape.dims)])
+    }
```
Shapes are supposed to be numbers, so maybe `Vec<Vec<usize>>` is better. Also, I don't think we need an option of vec; an empty vec will do the job if no shapes are provided.

syl20bnr commented:
Update the file name with a bench name and the first 8 chars of a random uuid, examples:
- `benchmarks_gelu_1b2bb391.json`
- `benchmarks_matmul_f79409f7.json`

The name is passed as an argument to the `save` function because each ran bench in a benchmark can have a different name and operation used. For instance with `gelu` where we have 3 batches of benches with different gelus.

louisfd commented:
> For the `.expect()` messages, I am not sure what to put in them. I see a lot of people putting an explanation of the error there and also sometimes people putting what is expected to work. The word "expect" suggests the latter but that's not what I see online most of the time 🤷🏻‍♂️ What's our code guideline on this ?

From [here](https://doc.rust-lang.org/std/result/enum.Result.html#recommended-message-style): _If you’re having trouble remembering how to phrase expect error messages remember to focus on the word “should” as in “env variable should be set by blah” or “the given binary should be available and executable by the current user”._

So I just stick to the word should and it works out

nathanielsimard commented on backend-comparison/src/persistence/base.rs:
```diff
@@ -1,100 +1,105 @@
+///      "operation": "operation name",
+///      "shapes": ["shape dimension", "shape dimension", ...],
+///      "timestamp": "timestamp",
+///      "numSamples": "number of samples",
```
`numSamples` => `num_samples`

syl20bnr replied:
See comment: https://github.com/tracel-ai/burn/pull/1131#discussion_r1448991987

nathanielsimard commented on backend-comparison/src/persistence/base.rs:
```diff
@@ -1,100 +1,105 @@
+///      "median": "duration in seconds",
+///      "mean": "duration in seconds",
+///      "variance": "duration in seconds"
+///      "rawDurations": ["duration 1", "duration 2", ...],
```
`rawDurations` => `raw_durations`

syl20bnr replied:
This is on purpose, I have a comment on it in the `seralize` function. This is the convention for MongoDB field names.

nathanielsimard commented on backend-comparison/src/persistence/base.rs:
```diff
@@ -1,100 +1,105 @@
-                .expect("Should have parsed to BenchmarkCommitResults struct"),
-            Err(_) => HashMap::default(),
-        };
+    let uuid = Uuid::new_v4().simple().to_string();
+    let short_uuid = &uuid[..8];
+    let file_name = format!("benchmarks_{}_{}.json", name, short_uuid);
+    let file_path = cache_dir.join(file_name);
```
I think using a timestamp is enough and will be better sorted on disk.

syl20bnr replied:
I can add it back, my reasoning was that we can already sort by creation date but that's not as simple indeed.

syl20bnr replied:
Oh and we can theoretically overwrite a file as I wrote initially.

Also it's handy to have the name  of the benchmark.

nathanielsimard replied:
yup keep the name, juste replacing the uuid with a timestamp, users should never run multiple benchmarks at the same time, the benchmarks won't be valid in this case anyways.

nathanielsimard commented on backend-comparison/src/persistence/base.rs:
```diff
@@ -1,100 +1,105 @@
+        // Serialize fields of BenchmarkResult
+        state.serialize_field("rawDurations", &self.results.raw.durations)?;
+        state.serialize_field("numSamples", &self.results.raw.durations.len())?;
+        state.serialize_field("mean", &self.results.computed.mean.as_secs_f64())?;
```
Serialize in the smallest time unit possible, since we can benchmark things that take microseconds.

louisfd commented on backend-comparison/benches/unary.rs:
```diff
@@ -17,6 +17,14 @@ impl<B: Backend, const D: usize> Benchmark for UnaryBenchmark<B, D> {
         "Unary Ops".into()
```
Name here

louisfd commented on backend-comparison/benches/unary.rs:
```diff
@@ -17,6 +17,14 @@ impl<B: Backend, const D: usize> Benchmark for UnaryBenchmark<B, D> {
     }
 
+    fn operation(&self) -> Option<String> {
+        Some("unary".to_string())
```
Name again here

louisfd commented on backend-comparison/benches/unary.rs:
```diff
@@ -41,7 +49,7 @@ fn bench<B: Backend>(device: &B::Device) {
     let benchmark = UnaryBenchmark::<B, D>::new(shape, num_repeats, device.clone());
 
-    Persistence::persist::<B>(vec![run_benchmark(benchmark)], device)
+    save::<B>("unary", vec![run_benchmark(benchmark)], device).unwrap();
```
Name again here

louisfd reviewed (changes requested):
It's very good in general. The only issue I see is how often we must tell the name of the benchmark (see my comments for example on unary). Could it not be unified in one place?

nathanielsimard reviewed (changes requested):
Instead of adding the new field "operation," we could add a field "ID" instead. "Name" is more generic, and we may have some benchmarks that aren't about single operations. Or maybe we need another structure:

- name (Could have a default implementation `core::any::type_name::<Self>().to_string()`)
- shapes
- options  (num_repeat, kind, etc.)

The ID could be generated from the concatenation of all of the above.

syl20bnr commented:
> It's very good in general. The only issue I see is how often we must tell the name of the benchmark (see my comments for example on unary). Could it not be unified in one place?

You are right, I thought about it but I cannot fetch the binary name at runtime, the only references are in Cargo,  in the command line, and the source file name. I'll try to find something.

syl20bnr commented:
Ready for another  round of review.
Updated gist: https://gist.github.com/syl20bnr/b0fdbbf0c3877b9e4ddf262a9cc388ac

Files on disk naming examples:

```
  bench_binary_1705007825641.json
  bench_from_data_1705007836909.json
  bench_gelu_reference_1705007872731.json
  bench_gelu_withcustomerf_1705007872987.json
  bench_gelu_withreferenceerf_1705007872929.json
  bench_matmul_1705007855627.json
  bench_to_data_1705007832378.json
  bench_unary_1705007818349.json
```

I added `num_repeats` also  to the documents,  I believe they can be useful.

nathanielsimard commented on backend-comparison/benches/custom_gelu.rs:
```diff
@@ -24,7 +24,16 @@ impl<B: Backend, const D: usize> Benchmark for CustomGeluBenchmark<B, D> {
 
     fn name(&self) -> String {
-        format!("Gelu {:?}", self.kind)
+        let kindstr = format!("{:?}", self.kind);
+        format!("gelu_{}", kindstr.to_lowercase())
```
This should be "gelu", the kind is included in the options

syl20bnr replied:
Ahah, ok I was expecting this one, I hesitated to do it. I did this to get a better filename but that's more a cosmetic tweak than a useful one.

nathanielsimard reviewed (changes requested):
I would not save the `num_repeat` in the database for now. I plan on eventually removing it since it is often confused with `num_samples`. It was useful when we didn't have a reliable way to sync a backend without actually reading the data. Repeating an execution meant that the data reading was less impactful.

nathanielsimard commented:
> I would not save the `num_repeat` in the database for now. I plan on eventually removing it since it is often confused with `num_samples`. It was useful when we didn't have a reliable way to sync a backend without actually reading the data. Repeating an execution meant that the data reading was less impactful.

@syl20bnr Though it might be a good idea to save the number of samples.

syl20bnr commented:
@nathanielsimard I already save the number of samples. See the gist.

I believe the repeat can still be useful, check the docstring I added to it, I think it is a good concept to have:

```
    /// Number of executions per sample
```

Say we have a very quick operation to bench, repeating it to get a meaningful sample can be useful.

syl20bnr commented:
Ideally the notion of repeat could be integrated to the main loop but I tried it and it was not as trivial to do.

syl20bnr commented:
We  decided to remove the num_repeat completely from the benchmakrs.

nathanielsimard reviewed (approved):
🎉

louisfd commented on burn-common/src/benchmark.rs:
```diff
@@ -12,27 +13,30 @@ use std::time::Instant;
-        *sorted.get(sorted.len() / 2).unwrap()
+        let min = *sorted.first().unwrap();
+        let max = *sorted.last().unwrap();
+        let var = *sorted.get(sorted.len() / 2).unwrap();
```
wrong variable name?

syl20bnr replied:
Nice catch ! Thank you !

louisfd commented on burn-common/src/benchmark.rs:
```diff
@@ -167,10 +224,84 @@ where
+            ],
+        };
+        let mean = durations.mean_duration();
+        assert_eq!(mean, Duration::from_secs(30));
```
That's a bit picky, but with these durations the mean is equal to the median. Maybe change them a bit so we can be sure we're finding the mean and not the median. You could remove the duration 50 and the mean should be 25.

louisfd reviewed (changes requested):
That's very nice, I think we reached the beautiful code stage. 
I have only two minor requests and then it's good to go
